src/sudokulogic.hs

```diff
@@ -0,0 +1,307 @@
+      cand = filter (\ x -> val `elem` getCandidates sud x)
+      isHorizontal = all (\ x -> fst x == fst pos) candidates
+      isVertical = all (\ x -> snd x == snd pos) candidates
+
+prop_lineBlock :: Bool
+prop_lineBlock = lineBlock sud (0,0) Two == [Horizontal] &&
+                 lineBlock sud (0,1) Five == [Horizontal] &&
+                 lineBlock sud (0,3) Six == [Horizontal]
+        where
+            sud = fillCell (fillCell (fillCell legalSudoku (0,0) Empty) (0,1) Empty) (0,2) Empty
+
+
+prop_singlePosition_section :: Position -> [SudVal]
+prop_singlePosition_section pos = getSinglePosition missingTopRowSud pos (blockPositions pos)
+
+
+-- | A Lemma That checks whether a cell is the only position in a section that can contain a value
+singlePositionSection :: Sudoku -> Position -> [Position] -> Bool
+singlePositionSection sud pos secPos = case getSinglePosition sud pos secPos of
+                                                [x] -> True
+                                                _ -> False
+
+
+getSinglePosition :: Sudoku -> Position -> [Position] -> [SudVal]
+getSinglePosition sud pos secPos  = filter (getOnlyCellInSectionFromValue sud pos secPos) [One .. Nine]
+
+getOnlyCellInSectionFromValue :: Sudoku -> Position -> [Position] -> SudVal -> Bool
+getOnlyCellInSectionFromValue sud pos secPos v = not (any (\x -> validPosition sud x v) emptyPos) && validPosition sud pos v
+    where
+        emptyPos = filter (\x -> not (isFilled $ valFromPos sud x)) secPos \\ [pos]
+
+
+validPosition :: Sudoku -> Position -> SudVal -> Bool
+validPosition sud pos v = v `elem` getCandidates sud pos
+
+valueInSection :: Section -> SudVal -> Bool
+valueInSection sec v = Filled v `elem` sec
+
+
+prop_lastCell :: Position -> Bool
+prop_lastCell (row, col) = lemmaLastCellInRow (fillCell illegalSudoku (row, col) Empty) (row, col) &&
+                           lemmaLastCellInColumn (fillCell illegalSudoku (row, col) Empty) (row, col) &&
+                           lemmaLastCellInBlock (fillCell illegalSudoku (row, col) Empty) (row, col)
+
+-- | Returns all possible candidates 
+getCandidates :: Sudoku -> Position -> [SudVal]
+getCandidates s pos = candidates
+    where
+        -- Takes all possible candidates and formats them to be SudVals
+        candidates = [x | x <- [One ..], Filled x `notElem` nub occupiedVals]
+
+        row = rowFromPos s pos
+        col = colFromPos s pos
+        block = blockFromPos s pos
+        -- 
+        lineCands = noteLineCandidatesRow (row \\ intersect row block) ++ noteLineCandidatesCol (col \\ (intersect col block))
+        
+        sections = [row, col, block]
+        
+        occupiedVals = values ++ map Filled pairs ++ map Filled lineCands
+
+        pairs =  nub (concatMap notePairs sections) \\ 
+                    case valFromPos s pos of
+                        Note xs -> getNotePairs xs
+                        _       -> []
+        values = nub $ concat sections
```

Comment:
You say: don't recompute candidates from scratch all the time, instead remember them in the data structure for the state and update them when new information comes in.

Also:
* Better to work with `SudVal`, not `Value`.
* If you work with sets instead of lists, then you don't have to call `nub` and checking for elementhood is much faster.
* Logically, this is quite a complex deduction. Can you split it up, for example using data structures in the state that track the intermediate steps?